kernel/src/syscall/signalfd.rs

```diff
@@ -237,25 +202,17 @@ impl SignalFile {
     }
 }
 
-impl Observer<SigEvents> for SignalFile {
-    // TODO: Fix signal notifications.
-    // Child processes do not inherit the parent's observer mechanism for signal event notifications.
-    // `sys_poll` with blocking mode gets stuck if the signal is received after polling.
-    fn on_events(&self, events: &SigEvents) {
-        if self
-            .signals_mask
-            .load(Ordering::Relaxed)
-            .contains(events.sig_num())
-        {
-            self.pollee.notify(IoEvents::IN);
-        }
-    }
-}
-
 impl Pollable for SignalFile {
     fn poll(&self, mask: IoEvents, poller: Option<&mut PollHandle>) -> IoEvents {
-        self.pollee
-            .poll_with(mask, poller, || self.check_io_events())
+        if let Some(poller) = poller {
+            let current = current_thread!();
+            current
+                .as_posix_thread()
+                .unwrap()
+                .register_signalfd_poller(poller, mask);
+        }
+
+        self.check_io_events() & mask
```

Comment:
It is better to avoid `current_thread!()` and `as_posix_thread()` in `check_io_events()`. It is only called here.